### ISSUE TITLE
Create subscription for new database

### ISSUE THREAD

**Issue by kashif** [nomadlabs/spacialdb#1]
Some references:
- https://stripe.com/docs/subscriptions
- https://github.com/peterkeen/payola

**Comment by sabman** [comment]
Also checkout https://github.com/RailsApps/rails-stripe-membership-saas/blob/master/db/schema.rb#L74

The app is using payola to build subscription based service

**Comment by lfcj** [comment]
@fraboeni and me are working on this issue together. We have a business logic question:
Context: user is already signed up, current URL: `http://localhost:3000/instances/new?plan=x` x->[1-4]
User picks a name for the instance/db and an available region. `CreateSubscription` gets fired up, in which a new sub is created. The next logical step is the payment process...or the droplet creation? 

Case 1: Payment process: We would then need to save the `region` and the `instanced_id` to :afterSuccessfulPayment created the doplet. 

Case 2: We call a `CreateDroplet` function before dealing with the payment, but then would need to save a droplet id to cancel/remove/destroy such droplet in case of payment failure.

**Comment by fraboeni** [comment]
And what kind of Image do you prefer for the droplets? Should it always be the same? Or is it changeable according to the plans/users?

**Comment by kashif** [comment]
yes its obviously will depend on the plan but for now lets not worry about it, you can just create a dummy instance entry on the payment going through. 

The way it works at digital ocean is that you have a spacialdb image with everyhting we want installed, then we use that image to create a droplet... but all this will be the task of the backend background task people...

**Comment by lfcj** [comment]
I have never worked with Github issues, and I am having trouble to know which answer is directed to which question. Contextually it seems you answered Franzi's...would you mind checking out my question?  (and maybe mentioning the person who asked the question in the answer?)

**Comment by kashif** [comment]
yes so the next step is the payment processing and if successful just create a dummy entry in the instances table

**Comment by fraboeni** [comment]
@sabman So should we use payola too, or stay with stripe as we talked about before?

**Comment by kashif** [comment]
well payola is using the stripe api, and i have tried to copy its style in the inital commits... you can use payola, but that would mean undoing the model and then adding payola migrations etc. 

but if you want to learn how the stripe api works etc. i would just use the payola code, and just try it with our model... 

but yes when we started payola did not have subscriptions... now it does... so perhaps it might be quicker to see how payola does it and copy it?

**Comment by kashif** [comment]
@lfcj perhaps we should add a `region_id` column to`instances`... since we have `region` as a table...

**Comment by lfcj** [comment]
@kashif that was my question. Thanks. I will proceed to change that in `schema.rb`.

**Comment by kashif** [comment]
@lfcj you don't change schema.rb, that gets generated automatically, you need to create migrations, see here: http://edgeguides.rubyonrails.org/active_record_migrations.html

**Comment by lfcj** [comment]
@kashif, ok, I have absolutely no experience with Ruby. Just googled migrations and changed `db/migrate/xxxxxx_create_instances.rb` by adding `t.integer :region_id`. Will that automatically add the column `region_id` the next time with `rake db:migrate`?

**Comment by kashif** [comment]
ah ok @lfcj it might be difficult to then do this if you have no experience with rails, perhaps take a day or so to go over: https://www.railstutorial.org/book 

rails is a very opinionated framework, do things the rails way and all is fine in the long run, do something strange and it bites you later on... so perhaps go over the tutorial to get the know the basics, espeically migrations and activerecord orm, which are one of the core features of rails and super helpful

**Comment by kashif** [comment]
@lfcj to add proper model associations, requires a bit more that the migrations, you also need to let the model know of the associations...

**Comment by lfcj** [comment]
@kashif I am doing my best to do tutorials on the fly, it would have of course been quite helpful to know what the project was going to be since March to acquire some preparation and not waste time out of already really short precious time...but I am doing what I can while also trying to understand the project. 

I understand the associations need to be changed, in order to create a relation between `region` and `instance`, therefore I have included a `belongs_to :region` in the Instance class. I originally thought of `has_one :region`, but since we will need to be able to change that (region) in the future, this [article](http://duanesbrain.blogspot.de/2006/05/ruby-on-rails-hasone-versus-belongsto.html) advises a `belongs_to`. The `instances` table will have a fk referring `regions`.

**Comment by kashif** [comment]
@lfcj sure the rails project was new after last week's discussions, since the inital plan as i said was to do certain small big data analytics usecases and blog about them... but it seems no one wants to do that... so i'm not sure what the best course of action is, since in order to do the spacialdb thing, or any other webapp project, you need to know the framework...

**Comment by fraboeni** [comment]
@kashif I guess it was not the point that we didn't want to. I can only talk for myself, but I was trying so hard all week long and I still felt like I didn't stand any chance of making it. I felt lost and there was not a lot of helf in the group. 
When you posted on Monday, that you would just mark the project as failed, I felt really frustrated because I felt like I did everything (even if I did not get to a solution, it was like 30-40 hours of my time)  for nothing. 
And so I was happy when I thought this new project might give us a chance of really doing something, and doing it in the group together. That's why I (and I guess its the same for others) voted for this project. 
But I can only agree that I would have liked it to be like your project last year (I talked to some people who participated and told me only the best) where people knew before what to do and how to prepare.

**Comment by lfcj** [comment]
@kashif we chose to change the project because no one had an idea of what the big data analytics project was about. There was not documentation, and just a bunch of tools (that could have been communicated in March) to learn without knowing exactly what for...The project has been there since March...5 months to prepare a minimum of documentation, tasks, ask for strengths, communicate tools to be used, etc...I will learn the framework, or any tools said, but that can not be done from one day to another (or in 2) and then implement flawlessly...

**Comment by kashif** [comment]
yes @fraboeni i felt the same too, but the projects we had talked to in march did not come through at the last minute, so i was left to improvise... which i understand is frustrating 

@lfcj so perhaps take this project as a way of learning rails, since with spacialdb we have no time dead-line and proceed at our own pace

**Comment by ghost** [comment]
@kashif you are absolutely right., mr. Kashif.
With 8 participiants, it s too less for making a big project.

**Comment by lfcj** [comment]
@sakonet the problem has not been the size of the project, but the lack of prior organization...but I guess that what's left is just keep working hard and try to make the best out of spacialdb.

**Comment by ghost** [comment]
@lfcj thats your opinion :-). I have no complains about the project.

**Comment by freaksN** [comment]
i like it so far too, we invested some time and maybe its little bit frustrating for someone but this investment was to learn something new. As @kashif said he was left to improvise so don't be mad and just relax :+1:

**Comment by ghost** [comment]
@freaksN :+1:

**Comment by ghost** [comment]
Now the best idea is that everyone just make familiarize with spacial db and finish the issues. I m Gettin closer with issue 1 and 2 to try to fix it

**Comment by lfcj** [comment]
@sakonet, we are also trying to work on issue 1 (under branch subscription), so far we created the subscription and have tried to include the region_id in instances by changing the associations and migrations. It would be cool if you could also share your ideas on 1  with us and help each other. Who are you on Slack?
